src/lib.rs

```diff
@@ -324,6 +348,40 @@ impl<T: Clone + Copy> RbProducer<T> for Producer<T> {
         self.data_available.notify_one();
         Some(cnt)
     }
+
+    fn write_blocking_timeout(&self, data: &[T], timeout: Duration) -> Result<Option<usize>> {
+        if data.is_empty() {
+            return Ok(None);
+        }
+        let guard = self.buf.lock().unwrap();
+        let mut buf = if self.inspector.is_full() {
+            let (guard, result) = self.slots_free.wait_timeout(guard, timeout).unwrap();
+            if result.timed_out() {
+                return Err(RbError::TimedOut);
+            }
+            guard
+        } else {
+            guard
+        };
```

Comment:
Both functions, `write_blocking` as well as `write_blocking_timeout` are almost identical except for this block.  What do you think of combining them, where `write_blocking` is a special case of `write_blocking_timeout` that just passes [`Duration::MAX`](https://docs.rs/time/latest/time/struct.Duration.html#associatedconstant.MAX)?  We could also fallback to

```rust
        let mut buf = if self.inspector.is_full() {
            self.slots_free.wait(guard).unwrap()
        } else {
            guard
};
```

in the case of `timeout` being `Duration::MAX` to keep the original behaviour.

Reply:
Good idea, thanks. Just committed this. I was thinking about moving the shared logic by all three methods into a helper function, but I figured that wasn't done originally to avoid code bloat caused by adding extra generic functions. This seems like a good way to reduce some of the duplication without creating any more generic functions than necessary.